Sources/Networking/NetworkController.swift

```diff
@@ -98,4 +98,17 @@ extension NetworkController: NetworkRequestPerformer {
             })
             .eraseToAnyPublisher()
     }
+    
+    public func send(_ request: any NetworkRequest, requestBehaviors: [RequestBehavior]) async throws -> NetworkResponse {
+        try await withUnsafeThrowingContinuation { continuation in
+            send(request) { result in
+                switch result {
+                case let .success(response):
+                    continuation.resume(returning: response)
+                case let .failure(failure):
+                    continuation.resume(throwing: failure)
+                }
+            }
+        }
+    }
```

Comment:
I think ultimately your approach here makes the most sense. I was thinking briefly that we may want to go async/await from top-to-bottom, but with the amount of code addition it would require, we'd just be further increasing the surface area of bugs with no real benefit as far as I can tell. It’d be fairly easy to mess up the handling of `requestBehaviors`, so relying on the `Result`-based approach here as you’ve done feels a lot safer to me.

---

[Please don’t use this suggestion] Since I typed it out while drafting this comment, I‘ll share, it would’ve looked something like:

```swift
public func send(_ request: any NetworkRequest, requestBehaviors: [RequestBehavior]) async throws -> NetworkResponse {
    let behaviors = defaultRequestBehaviors + requestBehaviors
    let urlRequest = makeFinalizedRequest(fromOriginalRequest: request.urlRequest, behaviors: behaviors)

    do {
        let (data, response) = try await urlSession.data(for: urlRequest)
        let networkResponse = try makeNetworkResponse(from: request, data: data, response: response)
        behaviors.requestDidFinish(result: .success(networkResponse))
        return networkResponse
    } catch {
        let networkError: NetworkError
        if let error = error as? NetworkError {
            networkError = error
        } else {
            networkError = NetworkError.underlyingNetworkingError(error)
        }
        behaviors.requestDidFinish(result: .failure(networkError))
        throw networkError
    }
}

private func makeNetworkResponse(from request: any NetworkRequest, data: Data?, response: URLResponse) throws -> NetworkResponse {
    if let statusCode = (response as? HTTPURLResponse)?.statusCode, !request.successHTTPStatusCodes.contains(statusCode: statusCode) {
        throw NetworkError.unsuccessfulStatusCode(statusCode: statusCode, data: data?.isEmpty == true ? nil : data)
    } else {
        return NetworkResponse(data: data, response: response)
    }
}
```

where `makeNetworkResponse` would then also be used by the combine implementation to reduce the duplication of determining the status code error.

Reply:
@ashlirankin18 please see this comment again. I’m saying we _shouldn’t_ use this more complex approach, and the approach you had here before was good.